x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/search/aggregations/support/ShapeValuesProvider.java

```diff
@@ -0,0 +1,38 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.spatial.search.aggregations.support;
+
+import org.apache.lucene.index.LeafReaderContext;
+import org.elasticsearch.index.fielddata.DocValueBits;
+import org.elasticsearch.index.fielddata.FieldData;
+import org.elasticsearch.index.fielddata.SortedBinaryDocValues;
+import org.elasticsearch.search.aggregations.support.ValuesSource;
+import org.elasticsearch.xcontent.ToXContentFragment;
+import org.elasticsearch.xpack.spatial.index.fielddata.ShapeValues;
+
+import java.io.IOException;
+
+public abstract class ShapeValuesProvider<T extends ToXContentFragment> extends ValuesSource {
+    public abstract ShapeValues<T> shapeValues(LeafReaderContext context);
```

Comment:
I wonder if the generic here should be ShapeValues<?> instead

Reply:
Addressed in the previous PR, now merged: https://github.com/elastic/elasticsearch/pull/89388